nix/third-party/gRPC-haskell/core/src/Network/GRPC/LowLevel/Client.hs

```diff
@@ -258,12 +258,14 @@ clientReader :: Client
              -> TimeoutSeconds
              -> ByteString -- ^ The body of the request
              -> MetadataMap -- ^ Metadata to send with the request
+             -> (ClientCall -> IO ())
```

Comment:
Is this the right API? Should you be returning the ClientCall?

Reply:
What do you suggest?

Reply:
Maybe we make a new type which only allow cancel to be called.
But returning the LL.ClientCall is more general.
And the user only gets the ClientCall if they explicitly ask for it.

Reply:
Returning the pair of ClientCall and the value would have been my approach, if it fits. If the only sensible thing you can do is cancel it then return a pair of `IO ()` (the cancel function) and the value. FWIW at a previous company I defined `newtype Cancel = Cancel (IO ())` and it was quite a successful abstraction.

Reply:
We can't return a pair because the interface is set up in a `withResource' style, the `ClientCall` value being obtained from `withClientCall`.

I was about to introduce the `Cancel` abstraction you suggested, when I noticed that `ClientCall` and `clientCallCancel` were already exposed in the `HighLevel` interface. (The only thing missing was how to obtain a `ClientCall` value.)

Since the type of `ClientCall` is abstract, the only thing to be done is call `clientCallCancel' on it. And so it seems to fulfill the role of the suggested newtype.

So I the end, didn't add it.

Reply:
Fair enough on all points.